### ISSUE TITLE
Document `av.open` and relationship to libavdevice

### ISSUE THREAD

**Issue by cefn** [mikeboers/PyAV#92]
Because the Pymedia project seems to be dead, I'm trying to identify maintained python libraries for audio recording, encoding and playback to be run on a Raspberry Pi. Sorry if this question doesn't count as an 'issue' and I'm happy to raise it elsewhere if that's more appropriate.

I've struggled to find reference material within the PyAV api reference which relates to opening and recording an audio stream from a device (e.g. a microphone) equivalent to [http://pymedia.org/tut/src/voice_recorder.py.html](http://pymedia.org/tut/src/voice_recorder.py.html)

In fact I'm finding it hard to find a reference to even av.open(...) is that because this python implementation mirrors some other API somewhere that I can refer to?

Is there any existing facility for recording from audio devices through PyAV, or would it have to be constructed from scratch? 

Does it make sense to wire, pyaudio, alsa or oss directly into PyAV? If so, which signatures in the API should I be looking at?
e.g. [Pulseaudio/Pyaudio](http://stackoverflow.com/questions/892199/detect-record-audio-in-python) [OSS](http://code.activestate.com/recipes/577587-record-and-playbac-using-standard-python/) [ALSA](http://stackoverflow.com/questions/1936828/how-get-sound-input-from-microphone-in-python-and-process-it-on-the-fly)

Any other suggested routes to achieving the same functionality with a maintained python library?

**Comment by mikeboers** [comment]
> Is there any existing facility for recording from audio devices through PyAV, or would it have to be constructed from scratch?

PyAV supports the underlying libavdevice, which allows for reading and writing from supported devices directly.

E.g., on my OS X laptop, the following (included example) will access the internal microphone:

~~~
python examples/decode.py --format avfoundation :0
~~~

> In fact I'm finding it hard to find a reference to even av.open(...) is that because this python implementation mirrors some other API somewhere that I can refer to?

It does not, and it is not documented. That is our bad, especially so for the first entrypoint. Ergo, this is being tagged as a documentation issue.

**Comment by cefn** [comment]
Thanks, @mikeboers 

Do you know what reference or tool can I use to understand the endpoints offered up by my specific libavdevice configuration? ```avfoundation``` is mac-specific so I'd have to investigate other options targeting Debian (raspbian) but I don't know where to begin.

If I'm right then the 'path' is actually ```:0``` (indicating the input from the first soundcard? where is this convention established? I need to get input from a bluetooth headset). ```avfoundation``` is a format identifier telling libav how to interpret the data which comes through when the ```:0```path is chosen, all handled in av.open() creating a decoding pipeline. The ```packet``` is actually a generator for frames of a given substream.  An ffplay process is lazily created on the first frame from a given stream's packet and decoded frames of bytes coming from the packet's generator is passed through to ffplay via STDIN to make sound if the "play" option is chosen (relying on the decode process having produced bytes which can be interpreted as raw wav).

I've tried to investigate open(...) in more detail, but at some point I guess I need to be looking at some libav documentation to find available options. Do you know where I can do this? I arrived at https://www.ffmpeg.org/doxygen/2.2/group__lavf__decoding.html#ga10a404346c646e4ab58f4ed798baca32 but no joy finding different values. Looks like the underlying header is https://www.ffmpeg.org/doxygen/2.2/avformat_8h_source.html but I don't know how to figure out how the names of valid formats which can be handled by an InputContainer. Any suggestions?

P.S. From my forensics so far I'd like to share that It's extremely hard to follow through all the extra non-functional code, and all the indirection which arises from argparser in decode.py, compared to an example invoked by python and not by the shell. Along the 'self-documenting' line I wonder if examples like decode.py could be separated into some minimally functional python (e.g. a Decode class with PyAV operations in it, which can be easily identified and traced, even maybe reused). A simple minimal python invocation could be offered with hard-coded values which do something. Separately an invocation could be shared with all the printed logging+argument handling to configure and trigger every part of the class's functionality. Argument parsing and logging dominates the code at present and obfuscates the access to PyAV so doesn't represent the goodness of the PyAV project as well as it might.

**Comment by mikeboers** [comment]
There are overall docs on [the FFmpeg site](https://www.ffmpeg.org/ffmpeg-devices.html), and `ffmpeg -devices` will tell you what is actually available on your machine.

You are correct about `"avfoundation"` being the "format", and `":0"` being the path to the media, which AVFoundation interprets as the first sound input. Unfortunately it is up to the libavdevice implementation to decide how to interpret it.

The rest of your commentary on documentation is pretty spot on. Something else that is very necessary is establishing strong links between the PyAV wrappers and the underlying library calls, so that we aren't responsible for documenting the entirety of what the underlying libraries can provide.

I welcome any patches regarding these matters.

**Comment by cefn** [comment]
For interest, this is what my Lubuntu environment reports from...
    ffmpeg -devices
This environment is a lot like Debian, though I'm waiting on the Raspberry Pi hardware right now. For now I'm just going to ignore the phrase "Output" which seems to apply to all my sound devices. I'm guessing at least some of them can be processed as inputs too. I think Alsa and OSS have lower resource demands than Pulse, but I'm not sure how else to distinguish between them before trying to get one to work.

```Devices:
 D. = Demuxing supported
 .E = Muxing supported
 --
 DE alsa            ALSA audio output
  E caca            caca (color ASCII art) output device
 D  dv1394          DV1394 A/V grab
 DE fbdev           Linux framebuffer
 D  jack            JACK Audio Connection Kit
 D  lavfi           Libavfilter virtual input device
 D  libcdio          
 D  libdc1394       dc1394 v.2 A/V grab
 D  openal          OpenAL audio capture device
  E opengl          OpenGL output
 DE oss             OSS (Open Sound System) playback
 DE pulse           Pulse audio output
  E sdl             SDL output device
  E v4l2            Video4Linux2 output device
 D  video4linux2,v4l2 Video4Linux2 device grab
 D  x11grab         X11 screen capture, using XCB
  E xv              XV (XVideo) output device```
